src/hardware/flash.rs

```diff
@@ -0,0 +1,89 @@
+use core::fmt::Write;
+use embedded_storage::nor_flash::{NorFlash, ReadNorFlash};
+use stm32h7xx_hal::flash::{LockedFlashBank, UnlockedFlashBank};
+
+#[derive(miniconf::Tree)]
+pub struct Settings {
+    pub broker: heapless::String<255>,
+    pub id: heapless::String<23>,
+}
+
+impl Settings {
+    fn new(mac: smoltcp_nal::smoltcp::wire::EthernetAddress) -> Self {
+        let mut id = heapless::String::new();
+        write!(&mut id, "{mac}").unwrap();
+
+        Self {
+            broker: "mqtt".into(),
+            id,
+        }
+    }
+}
+
+pub struct FlashSettings {
+    flash: LockedFlashBank,
+    pub settings: Settings,
+}
+
+impl FlashSettings {
+    pub fn new(
+        mut flash: LockedFlashBank,
+        mac: smoltcp_nal::smoltcp::wire::EthernetAddress,
+    ) -> Self {
+        let mut settings = Settings::new(mac);
+        let mut buffer = [0u8; 256];
+        let mut offset: usize = 0;
+
+        // We iteratively read the settings from flash to allow for easy expansion of the settings
+        // without losing data in the future when new fields are added.
+        flash.read(offset as u32, &mut buffer[..]).unwrap();
```

Comment:
This structure is kind of annoying and repetitive. We could use a proc-macro, but ideally we wouldn't need to and we could iterate over all the fields of the `settings` structure. However, the issue I have is that I need the `type` of the field as well to be able to deserialize it, and I don't think we can really iterate over types.